### ISSUE TITLE
[ID tool] The status of the grant should match the status of the grant from Grants.gov

### ISSUE THREAD

**Issue by mhuang-usdr** [usdigitalresponse/usdr-gost#52]

**Comment by mhuang-usdr** [comment]
This is dependent on #51.

**Comment by mhuang-usdr** [comment]
We should be syncing every 30 minutes.

Here's a test case (sorry, I should have added these at the beginning!)

On Grants.gov, I see `O-BJS-2022-171212 FY 2022 National Survey of Prosecutors` which was closed on 4/11/2022.

However, when I search for this grant in our tool, it still says that the status is `Posted`

<img width="636" alt="image" src="https://user-images.githubusercontent.com/91221265/167906180-325b4331-71d9-4976-9c04-825b27764a2d.png">

**Comment by mhuang-usdr** [comment]
Re-confirming that this is still an issue with the same test case as above

**Comment by chrisxkeith** [comment]
@mhuang-usdr : I searched for this grant in both production and stage, but don't see it. I am logged in as chris.keith@gmail.com/USDR tenant. Do I need to be logged into a different account/tenant? Thanks!
![image](https://user-images.githubusercontent.com/3921494/198105882-06af5d70-db14-4936-87bb-99daa08c1be9.png)
![image](https://user-images.githubusercontent.com/3921494/198106339-dc956d67-1602-415f-8070-b2195c83216c.png)

**Comment by mhuang-usdr** [comment]
It's because we had keywords filtering those results (I've made this mistake many times too...)

I deleted the keywords, try again! @chrisxkeith

**Comment by chrisxkeith** [comment]
After doing some software archeology (A.K.A. reverse engineering :), I think the problem is [here](https://github.com/usdigitalresponse/usdr-gost/blob/34c1b36abd75c8234d098e136aa30c4186124b4a/packages/server/src/lib/grantsgov.js#L126), where it looks like the server is only querying grants.gov within a date range going back 56 days (unless there is some environment variable set in render that changes this value). This means that this specific record will never be updated, as it is too far in the past.

Reading this code raises a number of questions for me. If anyone knows any answers, can you put them into this conversation?

1. Which date in the grants record from grants.gov is used to determine whether the grant falls within the 56 day range? Open date? Close date? Some date that isn’t returned by the grants.gov API?
2. What is significant about 56 days? It’s not enough for two months (62 days), or evenly divisible by 7 (so not an even number of weeks). Is there some limitation in the grants.gov site that determines this?
3. Why are the queries to grants.gov broken down into chunks of 10? Is this because of a limitation on the grants.gov site? Or some limitation in render.com?

Lastly, here are some options for ‘fixing’ it. I put fixing in quotes because I don’t know the intended functionality.

1. Do nothing, live with it.
2. Add an ‘info’ icon and some help text ‘under it’ on the grants modal that tells end users that they might want to verify the original record in grants.gov.
3. Expand the date range. If so, we’ll need to determine what a reasonable range would be. Note that this runs the risk of causing problems (see question 2 & 3 above).
4. A variation on this would be to set the date range to some number of months (6? 9?) on the weekend. After the scraper runs once, we can set the date range back to 56 (or whatever makes sense).

I may try some temporary code changes to see what the behavior is if I do increase that date range.

I plan to have (at least) one pull request for this issue that will put better, clearer information into the log. I don’t plan any pull requests (that will make functionality changes) until we can answer most of the questions above. Would be nice to get them all answered.

**Comment by chrisxkeith** [comment]
After doing some software archeology (A.K.A. reverse engineering :), I think the problem is [here](https://github.com/usdigitalresponse/usdr-gost/blob/34c1b36abd75c8234d098e136aa30c4186124b4a/packages/server/src/lib/grantsgov.js#L126), where it looks like the server is only querying grants.gov within a date range going back 7 days. Here is the environment setting in the production Render instance (same as in [.env.example](https://github.com/usdigitalresponse/usdr-gost/blob/34c1b36abd75c8234d098e136aa30c4186124b4a/packages/server/.env.example#L11)):
```
render@srv-...-mx666:~/project/src$ env | grep GRANTS_SCRAPER_DATE_RANGE
GRANTS_SCRAPER_DATE_RANGE=7
```
This means that this specific record will never be updated, as it is too far in the past (Posted Date: 4/5/2022, Close Date: 4/10/2022)

I don’t know whether the date range applies to the Posted Date or the Close Date. I would hope it’s Close Date. Does anyone know which one?

Here are two possible fixes:

1. Manually fix the value directly in the database. This is minimally invasive, and unlikely to cause any other problems. I plan to do this.
2. If anyone thinks that this is not a one-off problem, and that there could be other grants that aren’t synchronized with grants.gov, we could expand the date range. This, however, will cause the scraping to take a lot longer. Scraping for 7 days takes 5 minutes, so going back to 4/5/2022 would take 140 minutes (02:20 HH:MM). It takes this amount of time because there is a [delay between each query](https://github.com/usdigitalresponse/usdr-gost/blob/34c1b36abd75c8234d098e136aa30c4186124b4a/packages/server/.env.example#L12) to grants.gov, and the queries themselves probably take some time.. If we do this, we should do it as a one-shot effort on a weekend, where we increase the date range and reduce the scraping frequency. After the scraper runs once, we then set the date range and scraping frequency back to their original values..

If we want to do the second option, I would prefer to open a new issue for that work.
 
I do plan to have a pull request for this issue that will put better, clearer information into the log.

**Comment by chrisxkeith** [comment]
@TylerHendrickson @mhuang-usdr More research today leads me to not bothering with option 1 above. My new plan is:

- Submit a pull request (in a day or three) that will
-- Put more useful information into the server log
-- Set the grant sync interval to every 3 hours _(or is this in the environment somewhere?)_
- Once that deploys to _staging: _(need some help here)_
-- Set the environment variable `GRANTS_SCRAPER_DATE_RANGE=220` (This should make sure that the offending grant shown above will be synchronized).
-- Restart the Render instance
- Once the grant scraper has finished
--  Reset `GRANTS_SCRAPER_DATE_RANGE=7`
-- Restart Render instance
-- Reset grant sync interval back
-- Download log
-- Review contents of log

Once I have confidence that this works, I can do it to the production instance on a. weekend.

As to why/how this happened, I have a theory which I describe in [this new issue](https://github.com/usdigitalresponse/usdr-gost/issues/553).

**Comment by mhuang-usdr** [comment]
Thanks for digging into this, @chrisxkeith! Super helpful.

I'm pretty sure this is a persistent issue, not a one-off problem, unfortunately. So fix 1 won't work.

There are likely other ways to fix -

- @TylerHendrickson curious whether you think access to the grants.gov SOAP API would fix this?
- There is also [this XML extract](https://www.grants.gov/xml-extract.html) that is released by grants.gov once a day that could possibly be used instead, @chrisxkeith if you're able to dig into it. But rewriting the scraping system to use the XML extract instead would be a big lift.

Now that you've unearthed the real issue, I think overall there will need to be an architecture discussion to this question. If you had simply wanted to straightforward bug, you can also pick up a different issue while this is sorted out!

**Comment by chrisxkeith** [comment]
XML file does not have enough data (e.g., missing Eligibility Codes). I have archived the xml-download-and-read code here: https://docs.google.com/document/d/1lTpjNL94XGji3eGMSczPDu19eEPOo9KeVdN_7C3dSMs/edit?usp=sharing . Given that it may be some time before we get access to the SOAP API, I will start implementing option 2 from above.

**Comment by chrisxkeith** [comment]
Note that the filter feature ( https://github.com/usdigitalresponse/usdr-gost/issues/53) will require updating more fields than just the date fields. Probably another good reason to increase `GRANTS_SCRAPER_DATE_RANGE`

**Comment by chrisxkeith** [comment]
Proposal:
1. Set `GRANTS_SCRAPER_DATE_RANGE` to 120 days (e.g., 4 months). This will take a bit less than an hour, and will update somewhere around 900 grants. (Scraper output: `elapsed: 00:51:37, totalGrantsReturned: 2229, uniqueGrantsProcessed: 898`)
2. Set the [grants scraper cron job](https://github.com/usdigitalresponse/usdr-gost/blob/678c55c817d25e417d207a7eb573df550088ecc9/packages/server/src/index.js#L16) to run once a day at 1:30am Eastern Time.

**Comment by mhuang-usdr** [comment]
Waiting on larger ingestion architecture, deprioritizing for now
